src/Orleans.Pipeline.Client/OrleansPipe.cs

```diff
@@ -28,97 +28,99 @@ internal class OrleansPipe<TToServer, TFromServer>(
+            {
+                await _grain.Unsubscribe(_observer);
+                clusterClient.DeleteObjectReference<IOrleansPipeObserver<TFromServer>>(_observer);
+            }
+        }
+        catch (Exception ex)
         {
-            await _observerTask;
+            logger.LogError(ex, "An error occurred during Stopping");
         }
-        if (this is { _grain: not null, _observer: not null })
+        finally
         {
-            await _grain!.Unsubscribe(_observer);
+            _writerStoppingTokenSource.Dispose();
+            _observerStoppingTokenSource.Dispose();
         }
     }
+
     #endregion
 
     #region IOrleansPipeObserver<TFromServer>
-    public Task ReceiveMessage(PipeTransferItem<TFromServer> message)
+    public async Task ReceiveMessage(PipeTransferItem<TFromServer> message)
     {
         if (message is { Mode: TransferMode.Data, Item: not null })
         {
-            _reader.Writer.TryWrite(message.Item);
+            await _reader.Writer.WriteAsync(message.Item);
         }
-        return Task.CompletedTask;
     }
     #endregion
 
-    private async Task RunWriter()
+    private async Task RunWriter(CancellationToken cancellationToken)
     {
         try
         {
-            await foreach (var result in _writer.Reader.ReadAllAsync(_writerStoppingTokenSource.Token))
+            using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _writerStoppingTokenSource.Token);
+            await foreach (var result in _writer.Reader.ReadAllAsync(linkedCts.Token))
             {
                 var writeData = new PipeTransferItem<TToServer>
                 {
                     Mode = TransferMode.Data,
                     Item = result,
                     Metadata = _grain.GetPrimaryKeyString(),
                 };
-                await _grain!.Write(writeData);
+                await _grain.Write(writeData);
             }
         }
         catch (OperationCanceledException)
         {
             logger.LogInformation("Writer stopped");
         }
     }
-    private async Task RunObserver()
+    private async Task RunObserver(CancellationToken cancellationToken)
     {
         try
         {
-            while (!_observerStoppingTokenSource.Token.IsCancellationRequested)
+            using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _observerStoppingTokenSource.Token);
+            while (!linkedCts.Token.IsCancellationRequested)
             {
-                await Task.Delay(TimeSpan.FromSeconds(100), _observerStoppingTokenSource.Token);
-                if (this is { _grain: not null, _observer: not null })
-                {
-                    logger.LogInformation("Renewing subscription");
-                    var newObserver = clusterClient.CreateObjectReference<IOrleansPipeObserver<TFromServer>>(this);
-                    await _grain.RenewSubscription(_observer, newObserver);
```

Comment:
Do i see correct here, that i do not need to unsubscribe and resubscribe with a new reference? (just confirmation that it is so)

Reply:
@BernhardPollerspoeck yes because the reference will be the same as before, and the `ObserverManager` (used on the server/silo side) simply updates an observer reference if it already has one in its list.